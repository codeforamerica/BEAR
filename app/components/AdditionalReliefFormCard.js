/* eslint-disable react/no-unescaped-entities */
/* eslint-disable no-unused-vars */
// @flow
import React, { Component } from 'react';

import FormCard, {
  FormCardContent,
  FormCardFooter,
  FormCardHeader
} from './FormCard';
import ContinueButton from './ContinueButton';
import GoBackButton from './GoBackButton';

type Props = {
  additionalReliefOptions: AdditionalReliefOptions,
  onEligibilityOptionSelect: (string, boolean) => void,
  onOptionsConfirm: void => void,
  updateDate: void => void,
  onBack: void => void
};

export default class AdditionalReliefFormCard extends Component<Props> {
  onContinue = () => {
    const { onOptionsConfirm, updateDate } = this.props;
    onOptionsConfirm();
    updateDate();
  };

  onGoBack = () => {
    const { onBack } = this.props;
    onBack();
  };

  toggleUnder21Select = () => {
    const { onEligibilityOptionSelect, additionalReliefOptions } = this.props;
    onEligibilityOptionSelect(
      'subjectUnder21AtConviction',
      !additionalReliefOptions.subjectUnder21AtConviction
    );
  };

  render() {
    const { additionalReliefOptions } = this.props;
    return (
      <FormCard>
        <FormCardHeader>Additional relief</FormCardHeader>
        <FormCardContent>
          <label htmlFor="dismiss_under_21" className="checkbox">
            Dismiss convictions for people who were convicted of one of the
            above convictions at an age of 21 or younger
            <input
              type="checkbox"
              name="dismiss_under_21"
              id="dismiss_under_21"
              value="true"
              defaultChecked={
                additionalReliefOptions.subjectUnder21AtConviction
              }
              onChange={this.toggleUnder21Select}
            />
          </label>
        </FormCardContent>
        <FormCardFooter>
          <ContinueButton onContinue={this.onContinue} />
          <GoBackButton onGoBack={this.onGoBack} />
        </FormCardFooter>
      </FormCard>
    );
  }
}
