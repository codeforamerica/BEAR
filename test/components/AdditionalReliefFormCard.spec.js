import sinon from 'sinon';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import AdditionalReliefFormCard from '../../app/components/AdditionalReliefFormCard';

Enzyme.configure({ adapter: new Adapter() });
const sandbox = sinon.createSandbox();

function setup() {
  const options = { subjectUnder21AtConviction: true };

  const onOptionsConfirmSpy = sandbox.spy();
  const onOptionChangeSpy = sandbox.spy();
  const updateDateSpy = sandbox.spy();
  const onBackSpy = sandbox.spy();
  const component = mount(
    <AdditionalReliefFormCard
      additionalReliefOptions={options}
      onEligibilityOptionSelect={onOptionChangeSpy}
      onOptionsConfirm={onOptionsConfirmSpy}
      updateDate={updateDateSpy}
      onBack={onBackSpy}
    />
  );
  return {
    options,
    component,
    onOptionChangeSpy,
    onOptionsConfirmSpy,
    updateDateSpy,
    onBackSpy
  };
}
afterEach(() => {
  sandbox.restore();
});

describe('AdditionalReliefFormCard component', () => {
  describe('clicking the continue button', () => {
    it('should call onOptionsConfirm once', () => {
      const { component, onOptionsConfirmSpy } = setup();
      component.find('#continue').simulate('click');
      expect(onOptionsConfirmSpy.called).toBe(true);
      expect(onOptionsConfirmSpy.callCount).toEqual(1);
    });
    it('should call updateDate once', () => {
      const { component, updateDateSpy } = setup('path/to/file');
      component.find('#continue').simulate('click');
      expect(updateDateSpy.called).toBe(true);
      expect(updateDateSpy.callCount).toEqual(1);
    });
  });

  describe('clicking the go back button', () => {
    it('should call onBack once', () => {
      const { component, onBackSpy } = setup();
      component.find('#goback').simulate('click');
      expect(onBackSpy.called).toBe(true);
      expect(onBackSpy.callCount).toEqual(1);
    });
  });

  describe('clicking the checkbox for convictions that occurred when under 21', () => {
    it('should call onOptionChange with the correct arguments', () => {
      const { component, onOptionChangeSpy } = setup();
      expect(
        component.props().additionalReliefOptions.subjectUnder21AtConviction
      ).toEqual(true);
      component.find('#dismiss_under_21').simulate('change');
      expect(onOptionChangeSpy.called).toBe(true);
      expect(onOptionChangeSpy.callCount).toEqual(1);
      const { args } = onOptionChangeSpy.getCall(0);
      expect(args[0]).toEqual('subjectUnder21AtConviction');
      expect(args[1]).toEqual(false);
    });
  });

  it('should match exact snapshot', () => {
    const options = { subjectUnder21AtConviction: true };
    const component = (
      <div>
        <AdditionalReliefFormCard additionalReliefOptions={options} />
      </div>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
