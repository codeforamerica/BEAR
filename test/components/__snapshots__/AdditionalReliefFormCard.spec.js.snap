// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalReliefFormCard component should match exact snapshot 1`] = `
<div>
  <div
    className="form-card"
  >
    <header
      className="form-card__header"
    >
      <h1
        className="form-card__title"
      >
        Additional relief
        <p>
          Choose from the following options to expand relief.
        </p>
      </h1>
    </header>
    <div
      className="form-card__content"
    >
      <div
        className="form-group grid"
      >
        <div
          className="form-question grid__item width-two-thirds"
        >
          Dismiss all Prop 64 convictions if the conviction happened when the person was 21 or younger.
        </div>
        <label
          className="shift-fourth-fifths checkbox"
          htmlFor="true_subjectUnder21AtConviction"
        >
          <span
            className="sr-only"
          >
            Select to dismiss all Prop 64 convictions if the conviction happened when the person was 21 or younger.
          </span>
          <input
            defaultChecked={true}
            id="true_subjectUnder21AtConviction"
            name="subjectUnder21AtConviction"
            onChange={[Function]}
            type="checkbox"
            value={true}
          />
        </label>
      </div>
      <div
        className="form-group grid"
      >
        <div
          className="form-question grid__item width-two-thirds"
        >
          Dismiss convictions for people who are older than X:
          <div>
            <label
              className="sr-only"
              htmlFor="number-select"
            >
              Choose a minimum age to consider people eligible for dismissal.
            </label>
            <div
              className="select"
            >
              <select
                className="select__element"
                id="subjectAgeThreshold-select"
                onChange={[Function]}
                value={12}
              >
                <option
                  value={40}
                >
                  40
                </option>
                <option
                  value={41}
                >
                  41
                </option>
                <option
                  value={42}
                >
                  42
                </option>
                <option
                  value={43}
                >
                  43
                </option>
                <option
                  value={44}
                >
                  44
                </option>
                <option
                  value={45}
                >
                  45
                </option>
                <option
                  value={46}
                >
                  46
                </option>
                <option
                  value={47}
                >
                  47
                </option>
                <option
                  value={48}
                >
                  48
                </option>
                <option
                  value={49}
                >
                  49
                </option>
                <option
                  value={50}
                >
                  50
                </option>
                <option
                  value={51}
                >
                  51
                </option>
                <option
                  value={52}
                >
                  52
                </option>
                <option
                  value={53}
                >
                  53
                </option>
                <option
                  value={54}
                >
                  54
                </option>
                <option
                  value={55}
                >
                  55
                </option>
                <option
                  value={56}
                >
                  56
                </option>
                <option
                  value={57}
                >
                  57
                </option>
                <option
                  value={58}
                >
                  58
                </option>
                <option
                  value={59}
                >
                  59
                </option>
                <option
                  value={60}
                >
                  60
                </option>
                <option
                  value={61}
                >
                  61
                </option>
                <option
                  value={62}
                >
                  62
                </option>
                <option
                  value={63}
                >
                  63
                </option>
                <option
                  value={64}
                >
                  64
                </option>
                <option
                  value={65}
                >
                  65
                </option>
              </select>
            </div>
          </div>
        </div>
        <label
          className="shift-fourth-fifths checkbox"
          htmlFor="true_dismissOlderThanAgeThreshold"
        >
          <span
            className="sr-only"
          >
            Select to dismiss convictions for people who are older than specified age.
          </span>
          <input
            defaultChecked={true}
            id="true_dismissOlderThanAgeThreshold"
            name="dismissOlderThanAgeThreshold"
            onChange={[Function]}
            type="checkbox"
            value={true}
          />
        </label>
      </div>
      <div
        className="form-group grid"
      >
        <div
          className="form-question grid__item width-two-thirds"
        >
          Dismiss convictions that occurred more than X years ago:
          <div>
            <label
              className="sr-only"
              htmlFor="number-select"
            >
              Choose a minimum number of years since last conviction for dismissal.
            </label>
            <div
              className="select"
            >
              <select
                className="select__element"
                id="yearsSinceConvictionThreshold-select"
                onChange={[Function]}
                value={3}
              >
                <option
                  value={1}
                >
                  1
                </option>
                <option
                  value={2}
                >
                  2
                </option>
                <option
                  value={3}
                >
                  3
                </option>
                <option
                  value={4}
                >
                  4
                </option>
                <option
                  value={5}
                >
                  5
                </option>
                <option
                  value={6}
                >
                  6
                </option>
                <option
                  value={7}
                >
                  7
                </option>
                <option
                  value={8}
                >
                  8
                </option>
                <option
                  value={9}
                >
                  9
                </option>
                <option
                  value={10}
                >
                  10
                </option>
                <option
                  value={11}
                >
                  11
                </option>
                <option
                  value={12}
                >
                  12
                </option>
                <option
                  value={13}
                >
                  13
                </option>
                <option
                  value={14}
                >
                  14
                </option>
                <option
                  value={15}
                >
                  15
                </option>
              </select>
            </div>
          </div>
        </div>
        <label
          className="shift-fourth-fifths checkbox"
          htmlFor="true_dismissYearsSinceConvictionThreshold"
        >
          <span
            className="sr-only"
          >
            Dismiss convictions that occurred more than X years ago.
          </span>
          <input
            defaultChecked={true}
            id="true_dismissYearsSinceConvictionThreshold"
            name="dismissYearsSinceConvictionThreshold"
            onChange={[Function]}
            type="checkbox"
            value={true}
          />
        </label>
      </div>
      <div
        className="form-group grid"
      >
        <div
          className="form-question grid__item width-two-thirds"
        >
          Dismiss all HS 11357, HS 11358, HS 11359, or HS 11360 convictions if those are the only convictions on an individual's record.
        </div>
        <label
          className="shift-fourth-fifths checkbox"
          htmlFor="true_subjectHasOnlyProp64Charges"
        >
          <span
            className="sr-only"
          >
            Select to dismiss convictions for people who only have HS 11357, 11358, 11359, 11360 convictions on their record.
          </span>
          <input
            defaultChecked={true}
            id="true_subjectHasOnlyProp64Charges"
            name="subjectHasOnlyProp64Charges"
            onChange={[Function]}
            type="checkbox"
            value={true}
          />
        </label>
      </div>
    </div>
    <div
      className="form-card__footer"
    >
      <button
        className="button button--primary"
        id="continue"
        onClick={[Function]}
        type="button"
      >
        Continue 
        <i
          className="icon-arrow_forward"
        />
      </button>
      <button
        className="button"
        id="goback"
        onClick={[Function]}
        type="button"
      >
        Go back 
        <i
          className="icon-replay"
        />
      </button>
    </div>
  </div>
</div>
`;
